### PR TITLE
Migrate documentation from markdown to asciidoc #271

### LINKED ISSUE #271: Migrate documentation from markdown to asciidoc

### LINKED ISSUE #449: UserGuide.adoc: Update links to other asciidoc documents to use smart links
When exporting an `.adoc` file to `.html`, the hyperlinks that points to other asciidoc files are not changed from `.adoc` to `.html`. This will cause links to become broken.

Smart links in asciidoc documents should be utilized, so that the extensions are changed during generation.

**Solution:**
See:
http://asciidoctor.org/docs/user-manual/#navigating-between-source-files

(Alternative?)
For example, in UserGuide.adoc, it should be changed from:

`link:UserGuide.adoc[UserGuide.adoc]`

to

`<<UserGuide.adoc#>>`

### PR DESCRIPTION
**Pull request by limmlingg**
Fixes #271 
Fixes #449

### CODE REVIEW

File: README.adoc [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+* It is a Java sample application intended for students learning Software Engineering while using Java as the main programming language.
+* It is *written in OOP fashion*. It provides a *reasonably well-written* code example that is *significantly bigger* (around 6 KLoC)than what students usually write in beginner-level SE modules.
+* What's different from https://github.com/se-edu/addressbook-level3[level 3]:
+** A more sophisticated GUI that includes a list  panel and an in-built Browser.
+** More test cases, including automated GUI testing.
+** Support for _Build Automation_ using Gradle and for _Continuous Integration_ using Travis CI.
+
+== Site Map
+
+* <<docs/UserGuide#, User Guide>>
```
Comment (kychua) [inline README.adoc]: I think you'll have to add `ifdef::env-github,env-browser[:outfilesuffix: .adoc]`in the header, otherwise this links to UserGuide.*html* in GitHub instead of UserGuide.*adoc* (see [asciidoctor's guide](http://asciidoctor.org/docs/user-manual/#navigating-between-source-files)). Same thing for links in other docs.

You can also add `ifdef::env-github,env-browser[:relfileprefix: docs/]` so you can write `<<UserGuide#, User Guide>>` instead of `<<docs/UserGuide#, User Guide>>`.

**Comment by CanIHasReview** [comment]
# v1

@limmlingg submitted v1 for review.

* [v1 1/1] [Migrate documentation from markdown to asciidoc](https://github.com/se-edu/addressbook-level4/pull/452/commits/0448b30179824f57a38a6f003df929a3ce961620)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/cfc8ecc72a73a989ecc58427e6ddd1c564361408...0448b30179824f57a38a6f003df929a3ce961620))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/452/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: build.gradle [hunk 2]
```diff
@@ -196,7 +196,8 @@ asciidoctor {
                experimental: true,
                numbered: true,
                sectlinks: true,
-               idprefix: ''  // for compatibility with GitHub preview
+               idprefix: '',  // for compatibility with GitHub preview
+			   idseparator: '-'
```
Comment (yamgent) [inline build.gradle]: The indentation is off. :P

File: docs/AboutUs.adoc [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+== Project Team
+
+=== http://www.comp.nus.edu.sg/~damithch[Damith C. Rajapakse]
+image::DamithRajapakse.jpg[width="150", align="left"]
+
+*Role*: Project Advisor
+
+'''
+
+=== http://github.com/lejolly[Joshua Lee]
```
Comment (yamgent) [inline docs/AboutUs.adoc]: Each person now has a header number for the HTML output, which looks weird. Perhaps this can be normal text, but bolded?

Comment (yamgent) [inline docs/AboutUs.adoc]: Alternatively, you can disable the header numbers. Try `:numbered!:`

See: http://asciidoctor.org/docs/asciidoc-writers-guide/#section-titles, need to scroll down a bit to see.

File: docs/DeveloperGuide.adoc [hunk 4]
```diff
@@ -8,6 +8,7 @@ ifdef::env-github[]
 :tip-caption: :bulb:
 :note-caption: :information_source:
 endif::[]
+ifdef::env-github,env-browser[:outfilesuffix: .adoc]
```
Comment (yamgent) [inline docs/DeveloperGuide.adoc]: `DeveloperGuide.adoc` and `UserGuide.adoc` can be in separate commits:

```
Modify DeveloperGuide.adoc to use smart links

When linking to other asciidoctor documents, we use the extension .adoc.

The extensions remain unchanged even when exporting to HTML.

Let's fix this by using smart links, such that asciidoctor will change
to the correct extension .html when exporting to HTML.
```

---

```
Modify DeveloperGuide.adoc to use appendix styles
```

File: docs/AboutUs.adoc [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+:imagesDir: images
+
+We are a team based in the http://www.comp.nus.edu.sg[School of Computing, National University of Singapore].
+
+== Project Team
+
+=== http://www.comp.nus.edu.sg/~damithch[Damith C. Rajapakse]
+image::DamithRajapakse.jpg[width="150", align="left"]
+
+*Role*: Project Advisor
```
Comment (Zhiyuan-Amos) [inline docs/AboutUs.adoc]: remove the * *

**Comment by yamgent** [comment]
As discussed `README.adoc` doesn't seem to generate a `.html` version? We will need this, otherwise https://se-edu.github.io/addressbook-level4/ will not have a main page.

File: docs/LearningOutcomes.adoc [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+
+*Resources*
+
+* https://www.youtube.com/watch?v=Un80XoRT1ME[A three-minutes video] of designing architecture of and discovering component APIs for a Game of Tic-Tac-Toe.
+
+'''
+
+== Use Assertions `[LO-Assertions]`
+
+Note how the AddressBook app uses Java `assert`s to verify assumptions.
```
Comment (Zhiyuan-Amos) [inline docs/LearningOutcomes.adoc]: `assert` is buggy

File: docs/LearningOutcomes.adoc [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+
+* Add some more automated GUI tests.
+
+'''
+
+== Apply Design Patterns `[LO-DesignPatterns]`
+
+Here are some example design patterns used in the code base.
+
+* *Singleton Pattern* : link:../src/main/java/seedu/address/commons/core/EventsCenter.java[`EventsCenter.java`] is Singleton class. Its single instance can be accessed using the `EventsCenter.getInsance()` method.
```
Comment (Zhiyuan-Amos) [inline docs/LearningOutcomes.adoc]: instance spelling

File: docs/AboutUs.adoc (hunk 3 above)
Comment (pyokagan) [inline docs/AboutUs.adoc]: Seems like `:numbered!:` cannot be part of the document header. i.e. this doesn't work:
```diff
diff --git a/docs/AboutUs.adoc b/docs/AboutUs.adoc
index 64455b0..dec410f 100644
--- a/docs/AboutUs.adoc
+++ b/docs/AboutUs.adoc
@@ -1,6 +1,7 @@
 = About Us
 ifdef::env-github,env-browser[:outfilesuffix: .adoc]
 :imagesDir: images
+:numbered!:
 
 We are a team based in the http://www.comp.nus.edu.sg[School of Computing, National University of Singapore].
 
```
but this works:
```diff
diff --git a/docs/AboutUs.adoc b/docs/AboutUs.adoc
index 64455b0..52496c7 100644
--- a/docs/AboutUs.adoc
+++ b/docs/AboutUs.adoc
@@ -2,6 +2,8 @@
 ifdef::env-github,env-browser[:outfilesuffix: .adoc]
 :imagesDir: images
 
+:numbered!:
+
 We are a team based in the http://www.comp.nus.edu.sg[School of Computing, National University of Singapore].
 
 == Project Team
```

Comment (pyokagan) [inline docs/AboutUs.adoc]: > The problem is the attribute numbered: true in build.gradle. Removing this attribute should fix the problem.

btw I don't think it's a good idea to remove it from the `build.gradle` file. I expect the number of files needing numbered headers to greatly outnumber the number of files that do not need numbered headers, and so `numbered: true` is a good reasonable default.

Comment (limmlingg) [inline docs/AboutUs.adoc]: > Seems like `:numbered!:` cannot be part of the document header.

Thanks for pointing that out!

> I expect the number of files needing numbered headers to greatly outnumber the number of files that do not need numbered headers, and so numbered: true is a good reasonable default.

I think all files that need numbered headers already have it in the header. When I removed `numbered: true` I didn't notice any changes in any other file aside from `AboutUs.adoc`

**Comment by CanIHasReview** [comment]
# v2

@limmlingg submitted v2 for review.

* [v2 1/3] [Modify DeveloperGuide\.adoc to use smart links](https://github.com/se-edu/addressbook-level4/pull/452/commits/b9c863be51bcf0b989ae610bfe542db611fd10f2)
* [v2 2/3] [Modify DeveloperGuide\.adoc to use appendix styles and list continuation](https://github.com/se-edu/addressbook-level4/pull/452/commits/4b5b36a3712d7c6e697da102a95d3c06e226ebe7)
* [v2 3/3] [Migrate documentation from markdown to asciidoc](https://github.com/se-edu/addressbook-level4/pull/452/commits/8381bb4cfc4c71dd6fb86815db27e107f2144741)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/cfc8ecc72a73a989ecc58427e6ddd1c564361408...8381bb4cfc4c71dd6fb86815db27e107f2144741)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/4b6b3a219b3fe81f18ec27092ac7db79bc8811d2~1...4b6b3a219b3fe81f18ec27092ac7db79bc8811d2))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/452/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: docs/DeveloperGuide.adoc [hunk 8]
```diff
@@ -252,7 +253,7 @@ e.g. `seedu.address.logic.LogicManagerTest`
 
 === Headless GUI Testing
 
-Thanks to the https://github.com/TestFX/TestFX[TestFX] library we use, our GUI tests can be run in the _headless_ mode. In the headless mode, GUI tests do not show up on the screen. That means the developer can do other things on the Computer while the tests are running. See link:UsingGradle.md#running-tests[UsingGradle.md] to learn how to run tests in headless mode.
+Thanks to the https://github.com/TestFX/TestFX[TestFX] library we use, our GUI tests can be run in the _headless_ mode. In the headless mode, GUI tests do not show up on the screen. That means the developer can do other things on the Computer while the tests are running. See link:<<UsingGradle#running-tests, UsingGradle.adoc>> to learn how to run tests in headless mode.
```
Comment (yamgent) [inline docs/DeveloperGuide.adoc]: `... See link:<<UsingGradle#running-tests, UsingGradle.adoc>>...`

Remove the `link:`.

**Comment by yamgent** [comment]
``` 
*API* : link:../src/main/java/seedu/address/ui/Ui.java[`Ui.java`]
```
This link doesn't work. :(

**Comment by CanIHasReview** [comment]
# v3

@limmlingg submitted v3 for review.

* [v3 1/4] [Modify DeveloperGuide\.adoc to use smart links](https://github.com/se-edu/addressbook-level4/pull/452/commits/44072296483d8f335468be1c9590723cc7bc8088)
* [v3 2/4] [Modify DeveloperGuide\.adoc to use appendix styles and list continuation](https://github.com/se-edu/addressbook-level4/pull/452/commits/eaaf5cea8660c9069e11817a75d25ed962fd5880)
* [v3 3/4] [Migrate documentation from markdown to asciidoc](https://github.com/se-edu/addressbook-level4/pull/452/commits/35e22e5edd62db07126af48fb1ecba83eef57124)
* [v3 4/4] [Use absolute path for links to files in documentation](https://github.com/se-edu/addressbook-level4/pull/452/commits/8eaeaac320072ca0b12e5f526415439fa64c7572)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/cfc8ecc72a73a989ecc58427e6ddd1c564361408...8eaeaac320072ca0b12e5f526415439fa64c7572)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/compare/ee30ff60f6f35159eca49098fa610d8debfd8b33~1...ee30ff60f6f35159eca49098fa610d8debfd8b33))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/452/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: docs/AboutUs.adoc (hunk 3 above)
Comment (pyokagan) [inline docs/AboutUs.adoc]: > When I removed numbered: true I didn't notice any changes in any other file aside from AboutUs.adoc

That's because `:sectnums:` is an alias for `:numbered:` (c.f. https://github.com/asciidoctor/asciidoctor/issues/684)

<sup>(And annoyingly, it seems GitHub swallowed this comment thread again. Not visible from the PR page.</sup>

Comment (limmlingg) [inline docs/AboutUs.adoc]: Yup. I meant that I don't think removing `numbered: true` affects our documentation at all (except in this instance). I can put it back in if you think it's necessary though.

Comment (pyokagan) [inline docs/AboutUs.adoc]: Ah nevermind, I just realised that we need the sections to be numbered on GitHub as well.

Okay, it's better to remove it, but please document that in a separate commit.

**Comment by CanIHasReview** [comment]
# v4

@limmlingg submitted v4 for review.

* [v4 1/5] [Modify DeveloperGuide\.adoc to use smart links](https://github.com/se-edu/addressbook-level4/pull/452/commits/44072296483d8f335468be1c9590723cc7bc8088)
* [v4 2/5] [Modify DeveloperGuide\.adoc to use appendix styles and list continuation](https://github.com/se-edu/addressbook-level4/pull/452/commits/eaaf5cea8660c9069e11817a75d25ed962fd5880)
* [v4 3/5] [Migrate documentation from markdown to asciidoc](https://github.com/se-edu/addressbook-level4/pull/452/commits/578f827008c43d19db4860debeac2e7668c0703d)
* [v4 4/5] [Use absolute path for links to files in documentation](https://github.com/se-edu/addressbook-level4/pull/452/commits/4d5f46ca3e6efdf0e5cebdbcdbed8d5029b488ca)
* [v4 5/5] [Remove numbered: true in build\.gradle](https://github.com/se-edu/addressbook-level4/pull/452/commits/0a084006208d64a2b96b77bd6a59711ed8339284)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/cfc8ecc72a73a989ecc58427e6ddd1c564361408...0a084006208d64a2b96b77bd6a59711ed8339284)) ([:chart_with_upwards_trend: Interdiff between v3 and v4](https://github.com/se-edu/addressbook-level4/compare/ec91877f64ba5da5876cf1712bfa422743f8b50e~1...ec91877f64ba5da5876cf1712bfa422743f8b50e))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/452/4/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

File: build.gradle [hunk 9]
```diff
@@ -196,7 +196,8 @@ asciidoctor {
                experimental: true,
                numbered: true,
                sectlinks: true,
-               idprefix: ''  // for compatibility with GitHub preview
+               idprefix: '',  // for compatibility with GitHub preview
+               idseparator: '-'
```
Comment (pyokagan) [inline build.gradle]: What's the rationale for this change as well? Needs to be in its own commit with explanation.

**Comment by CanIHasReview** [comment]
# v5

@limmlingg submitted v5 for review.

* [v5 1/6] [Modify DeveloperGuide\.adoc to use smart links](https://github.com/se-edu/addressbook-level4/pull/452/commits/44072296483d8f335468be1c9590723cc7bc8088)
* [v5 2/6] [Add idseparator: '\-' attribute to build\.gradle](https://github.com/se-edu/addressbook-level4/pull/452/commits/16e7ada8773c884ac90320a1b0ac2266648f67ab)
* [v5 3/6] [Modify DeveloperGuide\.adoc to use appendix styles and list continuation](https://github.com/se-edu/addressbook-level4/pull/452/commits/3520462d11563c02dbd6af9d88de0eb02385c2fa)
* [v5 4/6] [Migrate documentation from markdown to asciidoc](https://github.com/se-edu/addressbook-level4/pull/452/commits/64b3abe4f3c180a37456b0310f69df376aa86794)
* [v5 5/6] [Use absolute path for links to files in documentation](https://github.com/se-edu/addressbook-level4/pull/452/commits/a766c5822f718ea8ea93fb12a19c22254aec2998)
* [v5 6/6] [Remove numbered: true in build\.gradle](https://github.com/se-edu/addressbook-level4/pull/452/commits/96cb99eb0f68adc8e50d1060a2e5b72a3a7d0d37)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/cfc8ecc72a73a989ecc58427e6ddd1c564361408...96cb99eb0f68adc8e50d1060a2e5b72a3a7d0d37)) ([:chart_with_upwards_trend: Interdiff between v4 and v5](https://github.com/se-edu/addressbook-level4/compare/1ca910dff5b56c78a005e896692cbcac6273039b~1...1ca910dff5b56c78a005e896692cbcac6273039b))
<details>
<summary>Checkout this PR version locally</summary>

```sh
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/452/5/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you wish to fetch this PR to.

</details>

**Comment by yamgent** [comment]
@kychua Do take a look if you have the time. :)
